molten-core/src/main/java/com/hotels/molten/core/collapser/FanOutRequestCollapser.java

```diff
@@ -198,29 +227,36 @@ private void logEmission(ContextWithValue<CONTEXT, VALUE> item) {
         }
     }
 
-    /**
-     * Creates a {@link FanOutRequestCollapser} builder over a bulk provider.
-     *
-     * @param <C>          the context type
-     * @param <V>          the VALUE type
-     * @param bulkProvider the bulk provider to delegate calls to
-     * @return the builder
-     */
-    public static <C, V> Builder<C, V> collapseCallsOver(Function<List<C>, Mono<List<V>>> bulkProvider) {
-        return new Builder<>(bulkProvider);
-    }
-
     private List<CONTEXT> toContexts(List<ContextWithSubject<CONTEXT, VALUE>> contextWithSubjects) {
         return contextWithSubjects.stream().map(cs -> cs.context).collect(Collectors.toList());
     }
 
-    private Mono<ContextWithValue<CONTEXT, VALUE>> bindValueToContext(ContextWithSubject<CONTEXT, VALUE> contextWithSubject, List<VALUE> values, String groupId) {
-        return Flux.fromIterable(values)
-            .filter(value -> matchContextByValue(contextWithSubject.context, value, groupId))
-            .next()
-            .map(value -> ContextWithValue.value(contextWithSubject, value))
-            .onErrorResume(e -> Mono.just(ContextWithValue.error(contextWithSubject, e)))
-            .defaultIfEmpty(ContextWithValue.empty(contextWithSubject));
+    private Mono<ContextWithValue<CONTEXT, VALUE>> bindValueToContext(List<ContextWithSubject<CONTEXT, VALUE>> contextsWithSubjects, VALUE value, String groupId) {
+        synchronized (contextsWithSubjects) {
+            ContextWithSubject<CONTEXT, VALUE> firstMatch = null;
+            for (var contextWithSubject : contextsWithSubjects) {
+                if (matchContextByValue(contextWithSubject.context, value, groupId)) {
+                    firstMatch = contextWithSubject;
```

Comment:
Iteration should be stopped once match is found.

Reply:
Good catch, fixed in 7f8a910. 🙂